database/rrdcontext.c

```diff
@@ -0,0 +1,2730 @@
+}
+
+static void rrdcontext_message_send_unsafe(RRDCONTEXT *rc, bool snapshot __maybe_unused, void *bundle __maybe_unused) {
+
+    // save it, so that we know the last version we sent to hub
+    rc->version = rc->hub.version = rrdcontext_get_next_version(rc);
+    rc->hub.id = string2str(rc->id);
+    rc->hub.title = string2str(rc->title);
+    rc->hub.units = string2str(rc->units);
+    rc->hub.family = string2str(rc->family);
+    rc->hub.chart_type = rrdset_type_name(rc->chart_type);
+    rc->hub.priority = rc->priority;
+    rc->hub.first_time_t = rc->first_time_t;
+    rc->hub.last_time_t = rrd_flag_is_collected(rc) ? 0 : rc->last_time_t;
+    rc->hub.deleted = (rc->flags & RRD_FLAG_DELETED) ? true : false;
+
+#ifdef ENABLE_ACLK
+    struct context_updated message = {
+        .id = rc->hub.id,
+        .version = rc->hub.version,
+        .title = rc->hub.title,
+        .units = rc->hub.units,
+        .family = rc->hub.family,
+        .chart_type = rc->hub.chart_type,
+        .priority = rc->hub.priority,
+        .first_entry = rc->hub.first_time_t,
+        .last_entry = rc->hub.last_time_t,
+        .deleted = rc->hub.deleted,
+    };
+
+    if(snapshot)
+        contexts_snapshot_add_ctx_update(bundle, &message);
+    else
+        contexts_updated_add_ctx_update(bundle, &message);
+#endif
+
+    // store it to SQL
+
+    if(rc->flags & RRD_FLAG_DELETED) {
+        rrdcontext_delete_from_sql_unsafe(rc);
+    }
+    else {
+        if (ctx_store_context(&rc->rrdhost->host_uuid, &rc->hub) != 0)
+            error("RRDCONTEXT: failed to save context '%s' version %lu to SQL.", rc->hub.id, rc->hub.version);
+    }
+}
+
+static bool check_if_cloud_version_changed_unsafe(RRDCONTEXT *rc, bool sending) {
+    bool id_changed = false,
+         title_changed = false,
+         units_changed = false,
+         family_changed = false,
+         chart_type_changed = false,
+         priority_changed = false,
+         first_time_changed = false,
+         last_time_changed = false,
+         deleted_changed = false;
+
+    if(unlikely(string2str(rc->id) != rc->hub.id))
+        id_changed = true;
+
+    if(unlikely(string2str(rc->title) != rc->hub.title))
+        title_changed = true;
+
+    if(unlikely(string2str(rc->units) != rc->hub.units))
+        units_changed = true;
+
+    if(unlikely(string2str(rc->family) != rc->hub.family))
+        family_changed = true;
+
+    if(unlikely(rrdset_type_name(rc->chart_type) != rc->hub.chart_type))
+        chart_type_changed = true;
+
+    if(unlikely(rc->priority != rc->hub.priority))
+        priority_changed = true;
+
+    if(unlikely((uint64_t)rc->first_time_t != rc->hub.first_time_t))
+        first_time_changed = true;
+
+    if(unlikely((uint64_t)(rrd_flag_is_collected(rc) ? 0 : rc->last_time_t) != rc->hub.last_time_t))
+        last_time_changed = true;
+
+    if(unlikely(((rc->flags & RRD_FLAG_DELETED) ? true : false) != rc->hub.deleted))
+        deleted_changed = true;
```

Comment:
Hi!. Should we check for the `rc->hub.*` not to be null to do the comparisons?

Reply:
if they are null, they changed. Do we dereference them?